docs/code_style.md

```diff
@@ -0,0 +1,226 @@
+	ChainConfig& config = kMainnetConfig;
+	ExecutionProcessor processor{block, *rule_set, buffer, *chain_config};
+
+Bad:
+
+	BlockNum expected_blocknum{previous_progress + 1};
+	ChainConfig& config{kMainnetConfig};
+	ExecutionProcessor processor(block, *rule_set, buffer, *chain_config);
+
+Copy initialization can use either style:
+
+	auto index_path{snapshot_path->index_file()};
+	auto index_path = snapshot_path->index_file();
+
+Exception:
+
+	Bytes transaction_key(8, 0);  // has to use parentheses to create 8 zeros instead of [8, 0] list
+
+### P1b init syntax convention for class members
+
+Member init list must use modern init syntax.
+
+	DebugExecutor(...) :
+	    database_reader_{database_reader},
+	    block_cache_{block_cache},
+	    workers_{workers},
+	    tx_{tx},
+	    config_{config} {}
+
+### P2 io_context var naming
+
+Use `ioc` name when having a single variable of type `asio::io_context` (unless having a more specific name brings much more expressiveness).
+
+Good:
+
+	io_context ioc;
+
+Bad:
+
+	io_context io_context;
+	io_context context;
+	io_context ctx;
+	io_context io;
+
+### P3 BlockNum var naming
+
+Use `BlockNum block_num` name when having a single variable of type `BlockNum` (unless having a more specific name brings expressiveness, e.g.: `start`, `end`, `last`). Do not use `height`.
+
+Good:
+
+	BlockNum block_num;
+
+Bad:
+
+	BlockNum height
+	BlockNum block_number
+	BlockNum number
+	BlockNum num
+	BlockNum bn
+	BlockNum b
+
+### P4 RAII lock type usage
+
+Use `std::scoped_lock` for a common case where a single mutex needs to be locked immediately and unlocked at the end of the scope. Use other lock types sparingly if other features are needed (e.g. multiple mutexes, deferred locking etc).
+
+### P5 SILKWORM_ASSERT
+
+Use `SILKWORM_ASSERT` instead of `assert(x)`.
+
+### P6 SILK_DEBUG logging macros
+
+Use `SILK_DEBUG` logging macros instead of `log::Debug()` syntax.
+
+### P7 explicit keyword
+
+Add explicit keyword, but only for single argument constructors, as per [CppCoreGuidelines C.46](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit)
+
+### P8 nodiscard
+
+Do not habitually use `[[nodiscard]]` by default. Use it sparingly as needed.
+
+We've had a "modernize-use-nodiscard" clang-tidy policy, which led to the proliferation of `[[nodiscard]]` throughout the codebase. The policy was disabled, but a lot of usages were redundant. For example, usage on Task is redundant, because the type itself is already marked with `[[nodiscard]]` within the library.
+
+An example where it is useful: if a function has a side effect, and returns no result or an error (e.g. `bool ok = f()`).
+
+### P9 move-only reference parameters
+
+Default to move by value. Use `T&&` parameters with caution.
+
+Legitimate uses are move-constructors. In some other cases it could be seen as a premature optimization. When copy-prevention is required, consider making a type move-only.
```

Comment:
IMO another legitimate usage is when `T` is an aggregate type. In fact, in such case the approach of forcing move by deleting the copy functions (i.e. copy-constructor and copy-assignment) defeats the aggregate semantics in the first place by asking you to provide a custom constructor.

Reply:
updated formulation